cpp/src/arrow/engine/logical_plan.h

```diff
@@ -0,0 +1,127 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```

Comment:
I'm fine with this file being somewhat WIP, a lot of stuff needs to be fleshed out and I expect refactoring. There's some other stuff that needs to be resolved either here or in follow up PRs